pkg/v1/partial/index.go

```diff
@@ -0,0 +1,53 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package partial
+
+import (
+	"fmt"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/match"
+)
+
+// FindManifests given a v1.ImageIndex, find the manifests that fit the matcher
+func FindManifests(index v1.ImageIndex, matcher match.Matcher) ([]v1.Descriptor, error) {
+	// get the actual manifest list
+	indexManifest, err := index.IndexManifest()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get raw index: %v", err)
+	}
+	manifests := []v1.Descriptor{}
+	// try to get the root of our image
+	for _, manifest := range indexManifest.Manifests {
+		if matcher(manifest) {
+			manifests = append(manifests, manifest)
+		}
+	}
+	return manifests, nil
+}
+
+// FindImages given a v1.ImageIndex, find the images that fit the matcher
+func FindImages(index v1.ImageIndex, matcher match.Matcher) ([]v1.Image, error) {
+	matches := []v1.Image{}
+	manifests, err := FindManifests(index, matcher)
+	if err != nil {
+		return nil, err
+	}
+	for _, desc := range manifests {
+		img, _ := index.Image(desc.Digest)
```

Comment:
We'll want to handle this error (I know this got copied from me 😄).

What should we do if a descriptor matches but points to a non-image artifact?

1. Let this error out.
2. Ignore non-Image media types.
3. Return an error for non-Image media types.

The second option makes me want to implement composition so we can:

```go
FindManifests(index, match.And(match.MediaTypes(types.Images), matcher))
```

But this might run afoul of the user intent -- they could just supply this matcher themselves! Should we make them? Or is that the whole point of this helper function?

Reply:
I am working really hard to avoid composition here. I don't object to it later, but I am focusing on simple; easy for the user to understand use cases, not to mention easy for me to grasp as we work these through.

Reply:
Let's think of it practically. I say, "give me all of the images from this index of platform X" or "give me all of the images from here of name Y". I don't really care about the fact that 7 of the 10 manifests do not point to images, I want those 3. The very calling of `FindImages` implies "it is images I want". I would ignore anything that errors because it is not an image.

As this is work-in-progress, I am going to add that handling in.

By the way, is there a `types.Images`? I didn't see that. Or were you suggesting it for the future.

Reply:
Updated.

Reply:
Updated

Reply:
> I am working really hard to avoid composition here. I don't object to it later, but I am focusing on simple; easy for the user to understand use cases, not to mention easy for me to grasp as we work these through.

Yeah that's fine, I don't want to commit to an API either. We could just hardcode some logic in here to check for the right media type (you could use a matcher to simplify) instead of using composition.

> I would ignore anything that errors because it is not an image.

I don't love this -- `index.Image(hash)` could fail for all kinds of reasons. Swallowing them feels dangerous.

> By the way, is there a types.Images? I didn't see that. Or were you suggesting it for the future.

Not yet, just theoretical shorthand for a group of roughly equivalent media types.

Reply:
> I don't love this -- index.Image(hash) could fail for all kinds of reasons. Swallowing them feels dangerous.

Try this; I just updated to use a custom error and check it with `errors.Is()`

Reply:
And also used that custom error in `v1/remote` and `v1/layout`